plugin/src/main/java/com/playmonumenta/redissync/RedisLock.java

```diff
@@ -0,0 +1,296 @@
+
+	private final Lock mInternalAccessLock;
+
+	private final Condition mIsIntraLockedCondition;
+	private final Condition mIsInterLockedCondition;
+
+	private volatile Optional<Thread> mLockOwner;
+
+	/**
+	 * Constructs a redis lock.
+	 *
+	 * @param lockName The key name postfix; acquiring
+	 * the inter-shard lock will set the key
+	 * {@code ConfigAPI.getServerDomain()}:locks:{@code lockName} in redis
+	 * to the name of this shard.
+	 * @param timeoutMS The expiry time for the key in redis, in milliseconds.
+	 * The expiry time of the key in redis is set to this with PX argument.
+	 * @param refreshIntervalMS While the lock is acquired, a separate thread
+	 * will refresh the expiry time on the key in redis every refreshIntervalMS
+	 * milliseconds.
+	 */
+	public RedisLock(String lockName, int timeoutMS, int refreshIntervalMS) {
+		mKeyName = ConfigAPI.getServerDomain() + ":locks:" + lockName;
+		mTimeoutMS = timeoutMS;
+
+		mSubscribeSignaler = Executors.newSingleThreadExecutor();
+
+		mRefreshScheduler = Executors.newSingleThreadScheduledExecutor();
+		mRefreshWorkers = Executors.newCachedThreadPool();
+		mRefreshIntervalMS = refreshIntervalMS;
+		mRefreshTask = Optional.empty();
+
+		mInternalAccessLock = new ReentrantLock();
+		mIsIntraLockedCondition = mInternalAccessLock.newCondition();
+		mIsInterLockedCondition = mInternalAccessLock.newCondition();
+
+		mLockOwner = Optional.empty();
+
+		RedisAPI.getInstance()
+				.pubSubConnection()
+				.addListener(new LockSubscriber());
+	}
+
+	private class LockSubscriber implements RedisPubSubListener<String, String> {
+
+		@Override
+		public void message(String channel, String message) {
+			if (!channel.equals("__keyspace@0__:" + mKeyName)) {
+				return;
+			}
+			mSubscribeSignaler.execute(() -> {
+				/**
+				 * Acquiring the lock here prevents signalling after
+				 * tryAcquireRedis fails but before condition await;
+				 * this prevents the thread from being dead until the next
+				 * expiry/deletion
+				 */
+				mInternalAccessLock.lock();
+				switch (message) {
+					case "del" -> mIsInterLockedCondition.signal();
+					case "expired" -> mIsInterLockedCondition.signal();
+					default -> { }
+				}
+				mInternalAccessLock.unlock();
+			});
+		}
+
+		@Override
+		public void message(String pattern, String channel, String message) {
+
+		}
+
+		@Override
+		public void subscribed(String channel, long count) {
+
+		}
+
+		@Override
+		public void psubscribed(String pattern, long count) {
+
+		}
+
+		@Override
+		public void unsubscribed(String channel, long count) {
+
+		}
+
+		@Override
+		public void punsubscribed(String pattern, long count) {
+
+		}
+
+	}
+
+	/**
+	 * Tries to acquire the lock on redis.
+	 * @return if acquiring the lock on redis was successful.
+	 */
+	private boolean tryAcquireRedis() {
+		return Optional.ofNullable(
+			RedisAPI.getInstance()
+				.sync()
+				.set(mKeyName, ConfigAPI.getShardName(), SetArgs.Builder.nx().px(mTimeoutMS))
```

Comment:
The shard name is not a sufficiently unique lock value for our uses. Multiple players may try to claim a lock on a market entry from the same shard, for instance. In the mailbox branch, I use a UUID, which will never conflict within our lifetimes.

Reply:
Since only one thread on the shard can attempt to claim the lock at a time due to the intra-lock, the shard name is sufficient.